### ISSUE TITLE
field alignments with prefixes

### ISSUE THREAD

**Issue by bdarcus** [bdarcus/bibtex-actions#94]
from @apc 

I've tried setting the value of `bibtex-actions-template-suffix` to include `${=has-pdf=:1}${=has-note=:1}`:. 

```
bibtex-actions-template-suffix is a variable defined in ‘bibtex-actions.el’.
Its value is
((t . "          ${=key=:15} ${=has-pdf=:1}${=has-note=:1}   ${=type=:12}    ${tags:*}"))
Original value was 
((t . "          ${=key=:15}    ${=type=:12}    ${tags:*}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry suffix.
When combined wiht the main template, the same string is used for
display and for search.
```

I've checked the value of `bibtex-actions-template` and it is what I assume it should be: 

```
bibtex-actions-template is a variable defined in ‘bibtex-actions.el’.
Its value is ((t . "${author:20}   ${title:48}   ${year:4}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry.
When combined with the suffix, the same string is used for
display and for search.
```

I now see the symbols indicating presence of notes and/or pdfs. But somehow they appear twice and break the alignment of the entries: 

<img width="1606" alt="image" src="https://user-images.githubusercontent.com/350504/114894458-a44fef00-9ddc-11eb-830c-49d7cbc28bea.png">

Any advice on how to adjust this? 

(Thanks again!)

I'm on Emacs 27.2.

_Originally posted by @apc in https://github.com/bdarcus/bibtex-actions/discussions/93_

**Comment by bdarcus** [comment]
This report from a user shows emacs 27.2 (a [Mac version](https://github.com/railwaycat/homebrew-emacsmacport)) attempting, incorrectly, to apply an affixation function, in the sense that content only defined in that function and only called from `affixation-function` is getting rendered in the display.

I thought `affixation-function` would simply be ignored for versions that don't support it, but not so.

Have you seen this before @minad?

Adding this as a check solves it for them:

``` elisp
                       (if (eq action 'metadata 'affixation-function)
                        (affixation-function . bibtex-actions--affixation))
```

Is that what I should be doing as a workaround?

Or should I consider this a bug in whatever version of Emacs they are using that I "wontfix"?

**Comment by apc** [comment]
According to [this issue](https://github.com/abo-abo/swiper/issues/2780), "Affixation functions work also in Emacs 27, only the default completion system will not use them." Not sure whether this is relevant or helpful or what.

**Comment by minad** [comment]
No, I have not seen something like this. I suggest you grep for `affixation-function` in the emacs source directory to make sure. And then you can take a look at which completion systems already use it (Selectrum, Vertico, maybe more?). This issue should be narrowed down a bit.

**Comment by bdarcus** [comment]
OK, thanks @minad.

Actually, my check turns it off for 28 too, so I don't want that.

**Comment by bdarcus** [comment]
@apc - do you think you can engage with the Mac Emacs project you are using, and see if they might be able to help sort it out on their end?

Or depending on how often they do releases, it might just get resolved in an update?

**Comment by apc** [comment]
I'll try contacting @railwaycat. Will be in touch if anything comes up. For now, I'll just stick with the fix you suggested earlier.

**Comment by apc** [comment]
Do you know what is the quickest way to test whatever is going on with `affixation-function` in my version of Emacs? Is there some quick test that doesn't require loading `bibtex-actions` and co? Thanks!

**Comment by apc** [comment]
OK, I just installed a different build from [here](https://emacsformacosx.com). But I'm running into the same issue: 

<img width="1283" alt="image" src="https://user-images.githubusercontent.com/350504/115096459-31856780-9ef3-11eb-80f2-b6d1924e2f47.png">

```
bibtex-actions-template-suffix is a variable defined in ‘bibtex-actions.el’.
Its value is
((t . "   ${=has-pdf=:1}${=has-note=:1} ${=key=:15}  ${=type=:12}    ${tags:*}"))
Original value was 
((t . "          ${=key=:15}    ${=type=:12}    ${tags:*}"))

  You can customize this variable.

Documentation:
Configures formatting for the BibTeX entry suffix.
When combined wiht the main template, the same string is used for
display and for search.
```

```
GNU Emacs 27.2 (build 1, x86_64-apple-darwin18.7.0, NS appkit-1671.60 Version 10.14.6 (Build 18G95)) of 2021-03-27
```

Not sure what to make of this...

**Comment by bdarcus** [comment]
> Do you know what is the quickest way to test whatever is going on with `affixation-function` in my version of Emacs? Is there some quick test that doesn't require loading `bibtex-actions` and co? Thanks!

Here's a minimal example, that AFAIK should only gives you a choice of A, B or C.

``` elisp
(defun test/affixation (cands)
  (mapcar
   (lambda (cand)
    (list cand "PRE " " SUFF")) cands))

(defun test/read ()
   (completing-read
    "choose one: "
    (lambda (string predicate action)
      (if (eq action 'metadata)
          `(metadata
	    (affixation-function . test/affixation))
	(complete-with-action action (list "A" "B" "C") string predicate)))))

(test/read)
```

On Emacs 28, you see the affixes though:

```
PRE A SUFF
PRE B SUFF
PRE C SUFF
```

To reproduce the alignment issue, I'm not sure.

**Comment by apc** [comment]
> > Do you know what is the quickest way to test whatever is going on with `affixation-function` in my version of Emacs? Is there some quick test that doesn't require loading `bibtex-actions` and co? Thanks!
> 
> Here's a minimal example, that AFAIK should only give you a choice of A, B or C on Emacs 27.
> 
> ```emacs-lisp
> (defun test/affixation (cands)
>   (mapcar
>    (lambda (cand)
>     (list cand "PRE " " SUFF")) cands))
> 
> (defun test/read ()
>    (completing-read
>     "choose one: "
>     (lambda (string predicate action)
>       (if (eq action 'metadata)
>           `(metadata
> 	    (affixation-function . test/affixation))
> 	(complete-with-action action (list "A" "B" "C") string predicate)))))
> 
> (test/read)
> ```
> 
> On Emacs 28, you see the affixes though:
> 
> ```
> PRE A SUFF
> PRE B SUFF
> PRE C SUFF
> ```
> 
> To reproduce the alignment issue, I'm not sure, but given what we've so far, we'd expect you to see some variation of this one.

FWIF, this is what I get: 

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/350504/115099599-060b7880-9f05-11eb-8e30-5965a380c894.png">

**Comment by bdarcus** [comment]
So affixation indeed works on your Emacs.

Did you alter the `bibtex-actions-symbols` defcustom at all?

The affixation function uses the cdr value of those entries, by default " ", when pdf or note is not present, which is what preserves alignment when using monospace fonts/symbols.

Which raises the second question: is your font monospace?

**Comment by apc** [comment]
I haven't changed the value of `bibtex-actions-symbols`, no. I am using the default value of `bibtex-completion-pdf-symbol` and of `bibtex-completion-notes-symbol`. `C-h v bibtex-actions-symbols` gives me this: 

```
bibtex-actions-symbols is a variable defined in ‘bibtex-actions.el’.
Its value is ((pdf "⌘" . #1=" ") (note "✎" . #1#) (link "🔗" . #1#))

  You can customize this variable.

Documentation:
Configuration alist specifying which symbol or icon to pick for a bib entry.
This leaves room for configurations where the absense of an item
may be indicated with the same icon but a different face.

[back]
```

For what is worth, I'm not getting worse misalignment issues with the notes symbol: 

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/350504/115114298-8a3d1a80-9f5c-11eb-8cd9-7527512e3268.png">

I am using a monospace font, yes (Input Mono). I'll try using `all-the-icons` as you suggest and report back. 

Thanks for your patience!

**Comment by bdarcus** [comment]
You also check what happens with vertico?

**Comment by apc** [comment]
Using `all-the-icons` works great, by the way: 

<img width="1273" alt="image" src="https://user-images.githubusercontent.com/350504/115115312-c161fa80-9f61-11eb-853f-efb46fabd476.png">

I'll look into `vertico`: should I disable `selectrum` for that? (See earlier comment on how I still don't understand how all these different packages interact with one another...)

**Comment by bdarcus** [comment]
I don't believe you need to deactivate selectrum; just turn on vertico-mode.

In any case, I often move back-and-forth between them while testing. You
don't need to change config or anything to do that.

**Comment by apc** [comment]
Here's another interesting fact: if I set the value of `bibtex-actions-symbols` from the `all-the-icons` example in the README so that at least *one* of the three values are `" "` if there's no corresponding note/file/link, the alignment works just fine: 

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/350504/115115751-1a329280-9f64-11eb-9efd-12ee84d0477a.png">

As soon as I have all three of them set to `" "`, though, the problem re-emerges: 

<img width="1274" alt="image" src="https://user-images.githubusercontent.com/350504/115115778-477f4080-9f64-11eb-8c6b-2df8642f955a.png">

Oh, and the same thing happens with `vertico`: 

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/350504/115115877-c2e0f200-9f64-11eb-8067-9d9e215a24d8.png">

**Comment by apc** [comment]
Just out of curiosity: alignment works alright for you if you use this? 

```
(setq bibtex-actions-symbols
  `((pdf . (,(all-the-icons-icon-for-file "foo.pdf" :face 'all-the-icons-dred) .
            " "))
    (note . (,(all-the-icons-icon-for-file "foo.txt") .
            " "))
    (link .
        (,(all-the-icons-faicon "external-link-square" :v-adjust 0.02 :face 'all-the-icons-dpurple) .
        " "))))
```

**Comment by bdarcus** [comment]
> if I set the value of `bibtex-actions-symbols` from the `all-the-icons` example in the README so that at least _one_ of the three values are `" "` if there's no corresponding note/file/link, the alignment works just fine... As soon as I have all three of them set to `" "`, though, the problem re-emerges:

I think that's because the link glyph is a different width than the others?

IIRC, that's why I used the "dim" face approach, rather than `" "`.

You haven't identified similar inconsistency with the default symbols; have you?

And what does `vertico` do with the default?

**Comment by apc** [comment]
> I think that's because the link glyph is a different width than the others?

Hmm. But that wouldn't make sense of the alignment being alright if I use `" "` for the link instead of the glyph but use the glyph for the pdf: 

<img width="1268" alt="image" src="https://user-images.githubusercontent.com/350504/115116422-6df2ab00-9f67-11eb-815e-90366f1b8cb6.png">

I'll check with the default symbols and see what `vertico` does with them as well. But first, I need help understanding the syntax of the values of `bibtex-action-template`. 

```
bibtex-actions-symbols is a variable defined in ‘bibtex-actions.el’.
Its value is ((pdf "⌘" . #1=" ") (note "✎" . #1#) (link "🔗" . #1#))

  You can customize this variable.

Documentation:
Configuration alist specifying which symbol or icon to pick for a bib entry.
This leaves room for configurations where the absense of an item
may be indicated with the same icon but a different face.

[back]

```

Trying to set the value to this: 

```
((pdf "⌘" . #1#) (note "✎" . #1#) (link "🔗" . #1#))

```

yields an error: `(invalid-read-syntax "#")`. 

I also tried something like this, just to see what would happen: 

```
(setq bibtex-actions-symbols '((pdf "⌘" . #1=" ") (note "✎" . "✎") (link "🔗" . #1#)))
```

but the result is not what I expected (I expected to see the notes glyph on all entries...). Any pointers?

**Comment by apc** [comment]
PS: using `all-the-icons` works the same way with `vertico`. If I use this: 

```
(setq bibtex-actions-symbols
  `((pdf . (,(all-the-icons-icon-for-file "foo.pdf" :face 'bibtex-actions-icon-on) .
            ,(all-the-icons-icon-for-file "foo.pdf" :face 'bibtex-actions-icon-dim)))
    (note . (,(all-the-icons-icon-for-file "foo.txt" :face 'bibtex-actions-icon-on) .
            " "))
    (link .
        (,(all-the-icons-faicon "external-link-square" :v-adjust 0.02 :face 'all-the-icons-dpurple) .
        " "))))
```

it works fine. If instead I use this: 

```
(setq bibtex-actions-symbols
  `((pdf . (,(all-the-icons-icon-for-file "foo.pdf" :face 'bibtex-actions-icon-on) .
            " "))
    (note . (,(all-the-icons-icon-for-file "foo.txt" :face 'bibtex-actions-icon-on) .
            " "))
    (link .
        (,(all-the-icons-faicon "external-link-square" :v-adjust 0.02 :face 'all-the-icons-dpurple) .
        " "))))
```

the entries appear misaligned.

**Comment by bdarcus** [comment]
> Its value is ((pdf "⌘" . #1=" ") (note "✎" . #1#) (link "🔗" . #1#))

Is that what the customization field looks like by default, or only after you made changes?

I never use that UI, so possible I did something wrong.

The `setq` examples you have demonstrate how the syntax should work, which is the value of that is a two-level alist; conceptually:

``` elisp
(pdf . ("P" . " "))
```

> PS: using all-the-icons works the same way with vertico.... It works fine. 

So let's call that a win; the instructions in the README work as expected, on your 27.2, and with selectrum and vertico.

As I said, I don't expect what you're trying to do to work, given my earlier experience with `all-the-icons`.

BTW, you can see the history for how we ended up here in #37; in particular this comment: https://github.com/bdarcus/bibtex-actions/pull/37#issuecomment-803191782.

But also, I was asking about the default symbols; not `all-the-icons`. That's the only remaining question in my view; why that happened to begin with. If it doesn't happen with vertico, it may be an issue with selectrum.

**Comment by bdarcus** [comment]
Here's what I see:

![Screenshot from 2021-04-17 10-57-00](https://user-images.githubusercontent.com/1134/115117299-d3e13180-9f6b-11eb-996d-8ac49d5c70d4.png)

**Comment by apc** [comment]
I don't use the customization interface either. The value I showed you is what I get when doing `C-h v` for `bibtex-actions-symbols`. I figured you were asking about the default, non-`all-the-icons` behavior. I just couldn't figure out how to understand the syntax used here:

```
bibtex-actions-symbols is a variable defined in ‘bibtex-actions.el’.
Its value is ((pdf "⌘" . #1=" ") (note "✎" . #1#) (link "🔗" . #1#))

  You can customize this variable.

Documentation:
Configuration alist specifying which symbol or icon to pick for a bib entry.
This leaves room for configurations where the absense of an item
may be indicated with the same icon but a different face.

[back]
```

(In particular, I don't know what the `#1" "` vs `#1#` distinction is supposed to be.) 

I'll try fiddling with the customization interface and see if I can get any clarity on what's happening with the default symbols.

**Comment by bdarcus** [comment]
Yeah that "#1" looks weird/wrong to me too!

**Comment by apc** [comment]
OK, messing with the customization interface allowed me to bypass trying to understand all of that. I now see this if I make one of the 'inactive' values to a non-empty glyph: 

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/350504/115117568-41da2880-9f6d-11eb-8439-4217939add23.png">

In contrast: 

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/350504/115117584-5cac9d00-9f6d-11eb-9140-f8c230ec86a0.png">

The exact same thing happens if I use `vertico` instead of `selectrum`.

¯\_(ツ)_/¯

**Comment by bdarcus** [comment]
That's odd. Even the first one, which makes zero sense, isn't perfectly aligned.

I think we can agree this isn't a bibtex-actions bug, and that you should use the `all-the-icons` solution?

BTW, you could use the same approach with the plain symbols; just need the "dim" face.

**Comment by apc** [comment]
> That's odd. Even the first one, which makes zero sense, isn't perfectly aligned.

How so? I don't see that...

> 
> I think we can agree this isn't a bibtex-actions bug, and that you should use the `all-the-icons` solution?
> 

Yes. Thanks for all your help! I'll close this issue. 

> BTW, you could use the same approach with the plain symbols; just need the "dim" face.

Good idea. Do you know if there's an easy way to just make the `dim` face always have as foreground the same color as its background? (So it's invisible even when in highlighted.)

**Comment by bdarcus** [comment]
> > BTW, you could use the same approach with the plain symbols; just need the "dim" face.
> 
> Good idea. Do you know if there's an easy way to just make the `dim` face always have as foreground the same color as its background? (So it's invisible even when in highlighted.)

No, I don't; why I didn't include it.

But if we could figure that out, could standardize on that approach and simplify the config.

**Comment by bdarcus** [comment]
> How so? I don't see that...

![Screenshot from 2021-04-17 15-07-55](https://user-images.githubusercontent.com/1134/115124150-c7220500-9f8e-11eb-9e59-6ac1e87cd624.png)

**Comment by apc** [comment]
> > How so? I don't see that...
> 
> ![Screenshot from 2021-04-17 15-07-55](https://user-images.githubusercontent.com/1134/115124150-c7220500-9f8e-11eb-9e59-6ac1e87cd624.png)

Whoa

**Comment by apc** [comment]
In case this helps anyone, here's what I ended up doing to avoid depending on `all-the-icons`: 

```
(use-package bibtex-actions
  :bind ("s-b" . bibtex-actions-open)
  :after embark
  :custom
  (bibtex-actions-template '((t . "${author:15}   ${title:40}   ${year:4}")))
  (bibtex-actions-template-suffix '((t . "   ${=key=:15}  ${=type=:12}    ${tags:*}")))
  (bibtex-completion-bibliography '("~/bib/apcmain.bib"))
  (bibtex-completion-pdf-field "File")
  (bibtex-completion-library-path "~/bib-library/") ; pdfs are assumed to be <key>.pdf
  (bibtex-completion-notes-path "~/org/notebook") ; notes are by default <key>.org
  :config
  (setf (alist-get 'bibtex embark-keymap-alist) 'bibtex-actions-map)
  (setq bibtex-actions-symbols
        `((pdf . (,(propertize "⌘" 'face `(:foreground ,(face-attribute 'default :foreground))) .
                  ,(propertize "⌘" 'face `(:foreground ,(face-attribute 'default :background)))))
          (note . (,(propertize "✎" 'face `(:foreground ,(face-attribute 'default :foreground))) .
                  ,(propertize "✎" 'face `(:foreground ,(face-attribute 'default :background)))))
          (link . (,(propertize "⤴" 'face `(:foreground ,(face-attribute 'default :foreground))) .
                  ,(propertize "⤴" 'face `(:foreground ,(face-attribute 'default :background))))))))
```

**Comment by apc** [comment]
OK, one more question, @bdarcus  (sorry!). Do you have any idea why the face properties specified as in the `all-the-icons` setup from the README are being overridden by `selectrum-mouse-highlight`? 

Perhaps it'd be best to explain the odd behavior I'm encountering. Using the values for `bibtex-actions-symbols` from that snippet in the README, I get (as expected) that the pdf icon (e.g.) for an entry without a corresponding pdf is displayed in the face `bibtex-actions-icon-dim`. If, however, my mouse happens to be hovering over that entry, the icon is displayed using (I think?) the `default` face: 

Without the pesky mouse hovering over anything: 

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/350504/115130990-4aa91980-9fc2-11eb-8a32-e542d3180553.png">

With the mouse: 

<img width="1273" alt="image" src="https://user-images.githubusercontent.com/350504/115130991-4d0b7380-9fc2-11eb-8aae-1a2b02d7b379.png">
